### ISSUE TITLE
Annotating callback argument as @tuplereturn

### ISSUE THREAD

**Issue by ElusiveMori** [TypeScriptToLua/TypeScriptToLua#811]
```ts
/** @vararg */
interface Vararg<T> extends Array<T> {}

/* Transpiled:
function vararger(...)
end
*/
function vararger(...args: Vararg<any>): void {}

/* Transpiled:
function hello(a, b, ...)
    vararger(...)
end
*/
function hello(a: number, b: number, ...rest: Vararg<any>): void {
    vararger(...rest)
}

/* Transpiled:
function ohno(callback)
    hello(
        1,
        2,
        table.unpack(
            callback()
        )
    )
end
*/
//                             V - ?
function ohno(callback: () => any[]) {
    hello(1, 2, ...callback())
}

/* Transpiled:
ohno(
    function() return {1, 2, 3} end
)
*/
ohno(() => [1, 2, 3])
```

When you have the opportunity, passing arguments directly by varargs is more efficient than packing/unpacking it. In LuaJIT especially, packing/unpacking is a massive pessimization that reduces performance to interpreter-level, whereas direct vararg passing is an optimized case (probably due to not allocating excess tables).

To this end, it would be nice to have a way to make the above code transpile into something like this instead:
```lua
function ohno(callback)
    hello(1, 2, callback())
end

ohno(function() return 1, 2, 3 end)
```

Maybe this could be the same `@tuplereturn` annotation but now also allowed on function parameters?

**Comment by ark120202** [comment]
It's already kind of supported: 
https://typescripttolua.github.io/play.html#src=%2F**%20%40vararg%20*%2F%0Ainterface%20Vararg%3CT%3E%20extends%20Array%3CT%3E%20%7B%7D%0A%0A%2F*%20Transpiled%3A%0Afunction%20vararger(...)%0Aend%0A*%2F%0Afunction%20vararger(...args%3A%20Vararg%3Cany%3E)%3A%20void%20%7B%7D%0A%0A%2F*%20Transpiled%3A%0Afunction%20hello(a%2C%20b%2C%20...)%0A%20%20%20%20vararger(...)%0Aend%0A*%2F%0Afunction%20hello(a%3A%20number%2C%20b%3A%20number%2C%20...rest%3A%20Vararg%3Cany%3E)%3A%20void%20%7B%0A%20%20%20%20vararger(...rest)%0A%7D%0A%0A%2F*%20Transpiled%3A%0Afunction%20ohno(callback)%0A%20%20%20%20hello(%0A%20%20%20%20%20%20%20%201%2C%0A%20%20%20%20%20%20%20%202%2C%0A%20%20%20%20%20%20%20%20table.unpack(%0A%20%20%20%20%20%20%20%20%20%20%20%20callback()%0A%20%20%20%20%20%20%20%20)%0A%20%20%20%20)%0Aend%0A*%2F%0A%0A%2F**%20%40tupleReturn%20*%2F%0Atype%20Callback%20%3D%20()%20%3D%3E%20any%5B%5D%3B%0Afunction%20ohno(callback%3A%20Callback)%20%7B%0A%20%20%20%20hello(1%2C%202%2C%20...callback())%0A%7D%0A%0Aohno(()%20%3D%3E%20%5B1%2C%202%2C%203%5D)%3B%0A

The problem is that `@tupleReturn` annotation has to be before declaration, so you can't just do `/** @tupleReturn */ () => any[]`. It would be nice to support, but it also would be simpler after #580.

**Comment by ElusiveMori** [comment]
Doesn't seem to work well with generics. #580 seems like it could offer a proper solution though.

[Here's an example with generics](https://typescripttolua.github.io/play.html#src=interface%20Event%3CC%20extends%20(...args%3A%20any%5B%5D)%20%3D%3E%20void%3E%20%7B%0A%20%20%20%20_1%3A%20C%0A%7D%0A%0A%2F**%20%40tuplereturn%20*%2F%0Atype%20Collector%3CC%20extends%20(...args%3A%20any%5B%5D)%20%3D%3E%20void%3E%20%3D%20()%20%3D%3E%20Parameters%3CC%3E%0A%0A%2F**%20%40tuplereturn%20*%2F%0Atype%20WorkingTuple%20%3D%20()%20%3D%3E%20any%5B%5D%0A%0Afunction%20makeEvent%3CC%20extends%20(...args%3A%20any%5B%5D)%20%3D%3E%20void%3E(%0A%20%20%20%20collector1%3A%20Collector%3CC%3E%2C%0A%20%20%20%20collector2%3A%20WorkingTuple%0A)%3A%20Event%3CC%3E%20%7B%0A%20%20%20%20let%20%5Ba%2C%20b%2C%20c%5D%20%3D%20((collector1%20as%20unknown)%20as%20()%20%3D%3E%20any%5B%5D)()%0A%20%20%20%20let%20%5Bd%2C%20e%2C%20f%5D%20%3D%20collector2()%0A%20%20%20%20return%20%7B%7D%20as%20Event%3CC%3E%0A%7D)

**Comment by ark120202** [comment]
It doesn't work because it's casted to a function without an annotation, [it works without the cast](https://typescripttolua.github.io/play.html#src=interface%20Event%3CC%20extends%20(...args%3A%20any%5B%5D)%20%3D%3E%20void%3E%20%7B%0A%20%20%20%20_1%3A%20C%0A%7D%0A%0A%2F**%20%40tuplereturn%20*%2F%0Atype%20Collector%3CC%20extends%20(...args%3A%20any%5B%5D)%20%3D%3E%20void%3E%20%3D%20()%20%3D%3E%20Parameters%3CC%3E%0A%0A%2F**%20%40tuplereturn%20*%2F%0Atype%20WorkingTuple%20%3D%20()%20%3D%3E%20any%5B%5D%0A%0Afunction%20makeEvent%3CC%20extends%20(...args%3A%20any%5B%5D)%20%3D%3E%20void%3E(%0A%20%20%20%20collector1%3A%20Collector%3CC%3E%2C%0A%20%20%20%20collector2%3A%20WorkingTuple%0A)%3A%20Event%3CC%3E%20%7B%0A%20%20%20%20let%20%5Ba%2C%20b%2C%20c%5D%20%3D%20collector1()%0A%20%20%20%20let%20%5Bd%2C%20e%2C%20f%5D%20%3D%20collector2()%0A%20%20%20%20return%20%7B%7D%20as%20Event%3CC%3E%0A%7D), or [with a cast to an annotated function](https://typescripttolua.github.io/play.html#src=interface%20Event%3CC%20extends%20(...args%3A%20any%5B%5D)%20%3D%3E%20void%3E%20%7B%0A%20%20%20%20_1%3A%20C%0A%7D%0A%0A%2F**%20%40tuplereturn%20*%2F%0Atype%20Collector%3CC%20extends%20(...args%3A%20any%5B%5D)%20%3D%3E%20void%3E%20%3D%20()%20%3D%3E%20Parameters%3CC%3E%0A%0A%2F**%20%40tuplereturn%20*%2F%0Atype%20WorkingTuple%20%3D%20()%20%3D%3E%20any%5B%5D%0A%0Afunction%20makeEvent%3CC%20extends%20(...args%3A%20any%5B%5D)%20%3D%3E%20void%3E(%0A%20%20%20%20collector1%3A%20Collector%3CC%3E%2C%0A%20%20%20%20collector2%3A%20WorkingTuple%0A)%3A%20Event%3CC%3E%20%7B%0A%20%20%20%20let%20%5Ba%2C%20b%2C%20c%5D%20%3D%20((collector1%20as%20unknown)%20as%20WorkingTuple)()%0A%20%20%20%20let%20%5Bd%2C%20e%2C%20f%5D%20%3D%20collector2()%0A%20%20%20%20return%20%7B%7D%20as%20Event%3CC%3E%0A%7D)

**Comment by ElusiveMori** [comment]
Oh, that makes sense now. That's actually a very useful trick.

With this, I'm able to get the behaviour I want now. Thanks. I think this can be closed, since there is already a way to make this work.
